### DISCUSSION TITLE
OCC Suggestions & feedback!

### DISCUSSION THREAD

**Discussion by btwise** [mackie100/OCClocalizations#108] [Ideas]
OC 0.6.0 added AppleFramebufferInfo in UEFI-->Protocol section

**Comment by mackie100** [comment]
> @mackie100
> 
> Should be add **gitee.com** alternate addresses to ocs.zip resource files and other resource file addresses that need to be downloaded. In some network environments, ocs.zip cannot be downloaded correctly

i'm trying to use this link https://gitee.com/btwise/OCClocalizations/raw/master/Resources/ocs/en/ocs.zip and
your gitee repository is causing a redirection, this is why cannot be downloaded correctly....

**Comment by mackie100** [comment]
> @mackie100
> In the third test version, the OC version is selected as 0.6.0 in the config checker, it will not call the corresponding 0.6.0 rule file, and the result is still 0.5.9!

Fixed, test 4th development
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/4870302/OCC.zip)

**Comment by mackie100** [comment]
> @mackie100
> 
> I have prepared another HTTP download address. If gitee fails to download the check, try to download the file from **http://disk.os-x.me:8800/OCC**. All resources are the same as github and there is no limit!

This doesn't work properly on this app... The response time is too long and the app hangs....
Can you change port from 8800 to 80?
Test 5th development
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/4875522/OCC.zip)

**Comment by mackie100** [comment]
> > > @mackie100
> > > I have prepared another HTTP download address. If gitee fails to download the check, try to download the file from **http://disk.os-x.me:8800/OCC**. All resources are the same as github and there is no limit!
> > 
> > 
> > This doesn't work properly on this app... The response time is too long and the app hangs....
> > Can you change port from 8800 to 80?
> > Test 5th development
> > [OCC.zip](https://github.com/mackie100/OCClocalizations/files/4875522/OCC.zip)
> 
> test OK!Please update the SMBIOS model and the latest fixed data!

SMBIOS MBP16,4 isn't yet complete...

**Comment by mackie100** [comment]
> @mackie100
> Clover Bootloader added Ocquirks, and the Clover Configurator should be updated!

The update is almost ready

**Comment by mackie100** [comment]
> > > @mackie100
> > > Clover Bootloader added Ocquirks, and the Clover Configurator should be updated!
> > 
> > 
> > The update is almost ready
> 
> CCG need add some new Localization files,OK?

CCG needs to check again all localization files.....

**Comment by mackie100** [comment]
@btwise @socialskyo test your localizations 
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/5045499/OCC.zip)

**Comment by socialskyo** [comment]
@mackie100 checked. So far so good. thank you.

**Comment by mackie100** [comment]
> @mackie100
> **in Misc-->Entries,when add a new entries,click `Browse` button,OCC will crash!Please fix it!**
> ![image](https://user-images.githubusercontent.com/38809560/91684375-e730bf80-eb89-11ea-90e1-cb277e80bfe0.png)

Crashlog?

**Comment by mackie100** [comment]
> > > **杂项中的**@ mackie100-
> > > **>条目，添加新条目时，单击`Browse`按钮，OCC将崩溃！请修复它！**
> > > ![图片](https://user-images.githubusercontent.com/38809560/91684375-e730bf80-eb89-11ea-90e1-cb277e80bfe0.png)
> > 
> > 
> > 崩溃日志？
> 
> `Process: OpenCore Configurator [844]
> Path: /Applications/OpenCore Configurator.app/Contents/MacOS/OpenCore Configurator
> Identifier: org.altervista.mackie100projects.OpenCore-Configurator
> Version: 2.11.1.1 (2.11.1.1)
> Code Type: X86-64 (Native)
> Parent Process: ??? [1]
> Responsible: OpenCore Configurator [844]
> User ID: 501
> 
> Date/Time: 2020-08-31 17:17:40.165 +0800
> OS Version: Mac OS X 10.14.6 (18G6020)
> Report Version: 12
> Anonymous UUID: 6BA59762-F50D-1CD8-F2FC-63127A271AF8
> 
> Time Awake Since Boot: 170 seconds
> 
> System Integrity Protection: disabled
> 
> Crashed Thread: 0 Dispatch queue: com.apple.main-thread
> 
> Exception Type: EXC_BAD_ACCESS (SIGSEGV)
> Exception Codes: KERN_INVALID_ADDRESS at 0x0000000000000000
> Exception Note: EXC_CORPSE_NOTIFY
> 
> Termination Signal: Segmentation fault: 11
> Termination Reason: Namespace SIGNAL, Code 0xb
> Terminating Process: exc handler [844]
> 
> VM Regions Near 0:
> -->
> __TEXT 000000010970d000-00000001098cf000 [ 1800K] r-x/r-x SM=COW /Applications/OpenCore Configurator.app/Contents/MacOS/OpenCore Configurator
> 
> Application Specific Information:
> Performing @selector(loadCustomEntries:) from sender NSButton 0x6000000f9970
> 
> Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
> 0 libsystem_platform.dylib 0x00007fff6713e6f2 _platform_strlen + 18
> 1 org.altervista.mackie100projects.OpenCore-Configurator 0x0000000109710a8b +[otherFunctions getUUID:] + 107
> 2 org.altervista.mackie100projects.OpenCore-Configurator 0x00000001097e7dc4 -[miscController loadCustomEntries:] + 884
> `

Check if crash occurs 
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/5163054/OCC.zip)

**Comment by mackie100** [comment]
> @mackie100
> In the kernel-> Force section, it should be allowed to drag and drop a kext file and automatically recognize the `BundlePath` information. Now, if I drag and drop a kext file, it will only be automatically added to the `Add` section. This is an error!

Isn't an error. Is not implemented yet

**Comment by mackie100** [comment]
> > Check if crash occurs
> > [OCC.zip](https://github.com/mackie100/OCClocalizations/files/5163054/OCC.zip)
> 
> The same problem has not been fixed!
> In the **`Misc-->boot-->Entries`**, the format of the `Path` is `PciRoot (0x0)/Pci (0x17, 0x0)/Sata (0x0, 0XFFFF, 0x0)/HD (1,GPT,BC6835C1-9507-43c4-30f5-D53A473D3F22, 0x800, 0x96000) / \EFI\boot\BOOTX64.EFI`, through `browse` button available?

Checking again...

**Comment by mackie100** [comment]
> @mackie100
> In the kernel-> Force section, it should be allowed to drag and drop a kext file and automatically recognize the `BundlePath` information. Now, if I drag and drop a kext file, it will only be automatically added to the `Add` section. This is an error!



> > Check if crash occurs
> > [OCC.zip](https://github.com/mackie100/OCClocalizations/files/5163054/OCC.zip)
> 
> The same problem has not been fixed!
> In the **`Misc-->boot-->Entries`**, the format of the `Path` is `PciRoot (0x0)/Pci (0x17, 0x0)/Sata (0x0, 0XFFFF, 0x0)/HD (1,GPT,BC6835C1-9507-43c4-30f5-D53A473D3F22, 0x800, 0x96000) / \EFI\boot\BOOTX64.EFI`, through `browse` button available?

Check again
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/5169607/OCC.zip)

**Comment by mackie100** [comment]
> > Check if crash occurs
> > [OCC.zip](https://github.com/mackie100/OCClocalizations/files/5163054/OCC.zip)
> 
> The same problem has not been fixed!
> In the **`Misc-->boot-->Entries`**, the format of the `Path` is `PciRoot (0x0)/Pci (0x17, 0x0)/Sata (0x0, 0XFFFF, 0x0)/HD (1,GPT,BC6835C1-9507-43c4-30f5-D53A473D3F22, 0x800, 0x96000) / \EFI\boot\BOOTX64.EFI`, through `browse` button available?

Yes, this value can be found in `nvram 4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102:boot-path`

**Comment by mackie100** [comment]
> > Check again
> > [OCC.zip](https://github.com/mackie100/OCClocalizations/files/5169607/OCC.zip)
> 
> again check,not been fixed!

Send me your IOREG file and this output: `nvram 4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102:boot-path`

**Comment by mackie100** [comment]
> > > > Check again
> > > > [OCC.zip](https://github.com/mackie100/OCClocalizations/files/5169607/OCC.zip)
> > > 
> > > 
> > > again check,not been fixed!
> > 
> > 
> > Send me your IOREG file and this output: `nvram 4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102:boot-path`
> 
> Does this work on your MacOS system?

Yes

**Comment by mackie100** [comment]
> I have WIN10,LINUX and MACOS three systems, NVRAM can only recognize WIN10 startup entries, how to display and select all custom entries?

OCC knows only OC boot-path, for other partitions do you need OC debug file or shell -> partitions scheme output

Test this. If OCC doesn't crash show me the alert window output.
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/5179933/OCC.zip)

**Comment by mackie100** [comment]
> @mackie100
> The 2.12.0 release seems a bit rushed; the `DisableLinkeditJettison` option should be in the **`kernel-->quirks`**, drivers add seem to be can't be dragged!

new version released now

**Comment by btwise** [comment]
@mackie100 
In occ2.12.1, no matter what language is selected in the preferences, ocs is the English rule for downloading. The ocs file of the corresponding language cannot be downloaded, and it cannot be downloaded from my OSX server.

**Comment by mackie100** [comment]
> @mackie100
> In occ2.12.1, no matter what language is selected in the preferences, ocs is the English rule for downloading. The ocs file of the corresponding language cannot be downloaded, and it cannot be downloaded from my OSX server.

fixing things........

**Comment by mackie100** [comment]
> @mackie100
> OCC 2.12.1 still has some logic problems:

1,2 points: For security reasons OCC works on EFI partition (config and other resources are located here)

What do you mean with "drag and drop function", drag for data reordering  or drag files?

**Comment by mackie100** [comment]
> > > @mackie100
> > > In occ2.12.1, no matter what language is selected in the preferences, ocs is the English rule for downloading. The ocs file of the corresponding language cannot be downloaded, and it cannot be downloaded from my OSX server.
> > 
> > 
> > fixing things........
> 
> When the preferred language is Chinese, the network resource address automatically takes precedence to **OSX**, if not accessible, then **GITEE**,**GITHUB**, if the user manually changes the resource address, then select priority by the user!
> 
> **GITEE** is not downloadable for downloading `ocs.zip` resources, so if the user switches the resource address to **GITEE**, the ocs.zip only will download from **OSX** resources by default, but other patch list resources, SMBIOS resources can still be obtained from **GITEE** address

What is the difference between gitee and osx?

**Comment by btwise** [comment]
> What is the difference between gitee and osx?

**GITEE** is our Chinese Git repository website, which is similar to **Github**, but anonymous users cannot download files exceeding 1MB, so they cannot download `ocs.zip`, but they can download other files such as patch resources and SMBIOS resource files. **OSX** is a private Git repository I built without any limited, but it is not open to the public and only serves as a private repository for you and me,So I mean, if  select **GITEE** in preferences, the `ocs.zip` is also downloaded from **OSX**, but other resources are still available from **GITEE**. Do I make myself clear?

**Comment by mackie100** [comment]
@btwise  Please test this 
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/5214613/OCC.zip)

**Comment by mackie100** [comment]
> > @btwise Please test this
> > [OCC.zip](https://github.com/mackie100/OCClocalizations/files/5214613/OCC.zip)
> 
> 1、In the ACPI-Add section, if the EFI partition is not mounted and the `browse` button is not prompted, I think any AML files outside of the EFI partition should be allowed to be added！We must consider that most of the time users edit config files in any folder, not within the EFI partition
> ![image](https://user-images.githubusercontent.com/38809560/93036635-0263fa80-f673-11ea-921a-7eab6afd9d5e.png)
> 2、In the Kernel-Add section,If EFI partition is not mounted, click the `browse` button again without any hint. I suggest adding a button corresponding to batch add EFI partition Kext and free add a single Kext file. Of course, there is no problem with dragging and adding, but is it reasonable to add one more button？
> ![image](https://user-images.githubusercontent.com/38809560/93037902-1f4dfd00-f676-11ea-824e-0c77f57b9b05.png)
> 3、For the **CONFIG CHECKER**, there is still no way to download the ocs.zip for the corresponding language, no matter what language you choose or the English OCS rules you download,Also, can you change the way you download packaged `ocs.zip` files to download rule files sequentially so that each file does not exceed 1MB and each of the three network addresses can be downloaded accordingly

1, 2 points: This works by drag and drop, added data after scanning EFI partitions or added an empty data by pressing plus button.

3. The OSX server causing a timeout, after 15s, OCC tries to download other resources...

Possible solution: change timeout from 15s to 60s

Test 
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/5226864/OCC.zip)

**Comment by btwise** [comment]
> 1. e OSX server causing a timeout, after 15s, OCC tries to download other resources...

Why did it time out? My repository Web access or download is normal. Would it be better to use SSL access? Here should be a hint and creat a log!

**Comment by btwise** [comment]
> The OSX server causing a timeout, after 15s, OCC tries to download other resources...
Possible solution: change timeout from 15s to 60s

**test ok!**

> 1, 2 points: This works by drag and drop, added data after scanning EFI partitions or added an empty data by pressing plus button.

When clicking the "scan/browse" button, if EFI partition is not mounted, a prompt will pop up, and then auto open the window to mount EFI!

**Comment by btwise** [comment]
@mackie100 
For the Kext loader, if Chinese is preferred, the kext loader cannot appear with a download list and the status bar prompts "Initialization: database cannot be read." No server found or Internet connection available."

The kext driver should also put the corresponding files into the `Resource` directory instead of downloading them from the Github author address, which is a pain for us
![image](https://user-images.githubusercontent.com/38809560/93300074-b78be380-f828-11ea-87a0-3c980b8b9c6b.png)

**Comment by mackie100** [comment]
> @mackie100
> For the Kext loader, if Chinese is preferred, the kext loader cannot appear with a download list and the status bar prompts "Initialization: database cannot be read." No server found or Internet connection available."
> 
> The kext driver should also put the corresponding files into the `Resource` directory instead of downloading them from the Github author address, which is a pain for us
> ![image](https://user-images.githubusercontent.com/38809560/93300074-b78be380-f828-11ea-87a0-3c980b8b9c6b.png)

Fixed. You can localize "category" key in kextsList database

**Comment by btwise** [comment]
> Fixed. You can localize "category" key in kextsList database

OK,can **kextdownloader** use `Localdirectory` don't use `Remotedirectory`？

**Comment by mackie100** [comment]
> > Fixed. You can localize "category" key in kextsList database
> 
> OK,can **kextdownloader** use `Localdirectory` don't use `Remotedirectory`？

No, remoteFolder is important key to get version of kext in a repository

**Comment by btwise** [comment]
> > > Fixed. You can localize "category" key in kextsList database
> > 
> > 
> > OK,can **kextdownloader** use `Localdirectory` don't use `Remotedirectory`？
> 
> No, remoteFolder is important key to get version of kext in a repository

Okay!

**Comment by btwise** [comment]
@mackie100 
For the outline mode, the current `LNPropertyListEditor` framework lacks the necessary drag and drop functions to find and replace the existing outline mode. Can OCC call python scripts to replace the current outline mode? **[ProperTree](https://github.com/corpnewt/ProperTree)**  is currently a popular program for outline mode editing, and can it be merged into OCC?
<img width="649" alt="OCC-PRO" src="https://user-images.githubusercontent.com/38809560/93575736-93690780-f9cc-11ea-9473-c3919df038dc.png">
<img width="649" alt="o" src="https://user-images.githubusercontent.com/38809560/93578084-7255e600-f9cf-11ea-9182-6fb5c63770ed.png">

**Comment by btwise** [comment]
@mackie100 
OC updated builtin firmware versions!
https://github.com/acidanthera/OpenCorePkg/commit/71952261a7587f1b27fd67ef12fec73dcc9fee42

**Comment by mackie100** [comment]
> @mackie100
> For the outline mode, the current `LNPropertyListEditor` framework lacks the necessary drag and drop functions to find and replace the existing outline mode. Can OCC call python scripts to replace the current outline mode? **[ProperTree](https://github.com/corpnewt/ProperTree)** is currently a popular program for outline mode editing, and can it be merged into OCC?
> <img alt="OCC-PRO" width="649" src="https://user-images.githubusercontent.com/38809560/93575736-93690780-f9cc-11ea-9473-c3919df038dc.png">
> <img alt="o" width="649" src="https://user-images.githubusercontent.com/38809560/93578084-7255e600-f9cf-11ea-9182-6fb5c63770ed.png">

this project can't merge python or propertree. I've already contacted the owner of that framework for implementing that features.....

**Comment by btwise** [comment]
> this project can't merge python or propertree. I've already contacted the owner of that framework for implementing that features.....

Look forward to your good results!

**Comment by btwise** [comment]
@mackie100 
OC added `MinKernel/MaxKernel` to Emulate and `LegacyCommpage` quirk …

**Comment by btwise** [comment]
@mackie100 
OCC2.15.0 will add the `Kernel-->Emulate- >DummyPowerManagement` keys copy to the `Kernel-Quirks`,please fix it!

**Comment by btwise** [comment]
@mackie100 
In the next update, change the `OSX` resource address from `http:// `to `https://` ,and add new resources address `https://codechina.csdn.net/btwise/occlocalizations `, named to `CSDN`,(for example, the download `ocs.zip` address is `https://codechina.csdn.net/btwise/occlocalizations/-/raw/master/Resources/ocs/zh-hans/ocs.zip`),other resource access addresses are similar, with **-/raw/master** added in the url!
**THX!**

**Comment by mackie100** [comment]
> @mackie100
> In the next update, change the `OSX` resource address from `http:// `to `https://` ,and add new resources address `https://codechina.csdn.net/btwise/occlocalizations `, named to `CSDN`,(for example, the download `ocs.zip` address is `https://codechina.csdn.net/btwise/occlocalizations/-/raw/master/Resources/ocs/zh-hans/ocs.zip`),other resource access addresses are similar, with **-/raw/master** added in the url!
> **THX!**

Please test resources networks changes
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/5359459/OCC.zip)

**Comment by btwise** [comment]
> > @mackie100
> > In the next update, change the `OSX` resource address from `http:// `to `https://` ,and add new resources address `https://codechina.csdn.net/btwise/occlocalizations `, named to `CSDN`,(for example, the download `ocs.zip` address is `https://codechina.csdn.net/btwise/occlocalizations/-/raw/master/Resources/ocs/zh-hans/ocs.zip`),other resource access addresses are similar, with **-/raw/master** added in the url!
> > **THX!**
> 
> Please test resources networks changes
> [OCC.zip](https://github.com/mackie100/OCClocalizations/files/5359459/OCC.zip)

The test has no problem, the address list can be unified into capital letters, when the user chooses Chinese language, the resource address had better be automatically set to "CSDN", because it is very fast for us!

**Comment by btwise** [comment]
@mackie100 
If the oc-config file is opened with the Clover configurator or the clover-config file is opened with the Opencore configurator, the warning information should be automatically recognized and given, and then the file should be closed automatically without any changes to the file. Although this is a small problem, sometimes double-clicking the config file will open the error configurator,so add this functions!

**Comment by btwise** [comment]
@mackie100 
please update SMBIOS database!
https://github.com/acidanthera/OpenCorePkg/commit/fd8c076a71131e6ea6b935a8a9288880b1b289e7

**Comment by btwise** [comment]
@mackie100 
OC added `CustomMemory` functions!please update occ!

**Comment by btwise** [comment]
@mackie100 
When the `Custommemory` is `false`, the controls in the `Memory `section should be completely unavailable and no config.plist is written. Only when the `Custommemory` is `true` will it be valid and written to config.plist to keep the configuration file clean!
Custom memory part of the content, it is best to provide automatic calculation function, let the user choose with the mouse, too complex professional knowledge, most users do not have, thus will input wrong data resulting in the system cannot start！

**Comment by btwise** [comment]
@mackie100 
OC added `ForceResolution` options in `UEFI->OUTPUT`

**Comment by btwise** [comment]
@mackie100 
https://github.com/acidanthera/OpenCorePkg/commit/39c1e1cc8a305089c205bd28e33559673d65e6ac

**Comment by btwise** [comment]
@mackie100 OC Release already to 0.6.3,and develop version to 0.6.4,please update

add `ForceSecureBootScheme` quirks on **kernel->quirks**
add `BlacklistAppleUpdate` quirks on **MIsc**

**Comment by btwise** [comment]
@mackie100 
please add `ForceSecureBootScheme` quirks on **kernel->quirks**!!

**Comment by btwise** [comment]
@mackie100 
![image](https://user-images.githubusercontent.com/38809560/98487261-f6547d80-225c-11eb-813d-eb7d4ef4872f.png)

**What's going on here?**

**Comment by mackie100** [comment]
> @mackie100
> ![image](https://user-images.githubusercontent.com/38809560/98487261-f6547d80-225c-11eb-813d-eb7d4ef4872f.png)
> 
> **What's going on here?**

Already fixed in next update. Auto layout issue.

**Comment by btwise** [comment]
@mackie100 
OC added new keys:
`Booter-->patch` array
`UEFI-->PlayChime` Change from a **Boolean** value to a **character** value

**Comment by btwise** [comment]
@mackie100 
`PickerAttributes` should add calculator functionality

**Comment by mackie100** [comment]
> @mackie100
> `PickerAttributes` should add calculator functionality

PickerAttributes has already a calculator

**Comment by btwise** [comment]
> > @mackie100
> > `PickerAttributes` should add calculator functionality
> 
> PickerAttributes has already a calculator

Sorry, `ConsoleAttributes` is missing a calculator

The optional value of the `PickerAttributes` calculator missing  a mouse-pointer control, as if it were 17？

**Comment by mackie100** [comment]
@btwise Please test OpenCore Validator tool 
[OCC.zip](https://github.com/mackie100/OCClocalizations/files/5777294/OCC.zip)
